### PR TITLE
Notify Slack of open Zendesk tickets every day

### PR DESCRIPTION
**Pull request by islemaster**
([LP-205](https://codedotorg.atlassian.net/browse/LP-205)) We recently reflected that engineers don't get a lot of ambient awareness of our Zendesk ticket volume - at least, I personally never have a sense of this.  I'm trying to address that with a lightweight daily notification to Slack.

image

At 10:00am every weekday this script will query the Zendesk API for relevant new/open tickets and post a count to Slack.  For now, this is limited to:

- Posts "Learning Platform" tickets to the  `#learning-platform` channel
- Posts "*Labs" tickets to the `#star-labs-cabal` channel

But it's ready to enable for other channels too. The message links to the appropriate view in Zendesk, and also to the script's own source to make it easy to find and update in the future.

**API rate limits.** This script makes 1+`n` API calls each time it runs, where `n` is the number of pages of search results we need to return.  It looks like we get about 100 tickets per page.  In testing today, that meant a maximum of 2 pages (3 calls total) when retrieving all configured groups at once.  Our rate limit with Zendesk is in the hundreds-per-minute, so this barely makes a dent.

### CODE REVIEW

File: bin/cron/zendesk_slack_report [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+# Slack room -> Zendesk group mappings to report on
+MONITORING_GROUPS = {
+  # "slack-room-name" => ["Zendesk group name", "Zendesk group name", ...]
+  "learning-platform" => ["Learning Platform"],
+  # "developers" => ["Developers"],
+  # "plc-tools" => ["PLC Tools"],
+  # "privacy" => ["Privacy"],
+  # "star-labs-cabal" => ["*Labs"],
+  # "user-accounts" => ["Accounts"],
```
Comment (islemaster) [inline bin/cron/zendesk_slack_report]: This script is all set up to enable notifications for several other channel/group combinations.  I plan to send an example notification and see who's interested.

File: bin/cron/zendesk_slack_report [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+# Zendesk group -> Zendesk filter id mappings for helpful links (optional)
+# rubocop:disable Style/NumericLiterals
+GROUP_FILTERS = {
+  "Developers" => 47222126,
+  "Learning Platform" => 360082967212,
+  "PLC Tools" => 114103940251,
+  "*Labs" => 301129028,
```
Comment (islemaster) [inline bin/cron/zendesk_slack_report]: The filter ids here override the Zendesk link posted for a particular group.  If one is not found here (e.g. "Accounts") we will link to the generic Zendesk search page instead.

File: bin/cron/zendesk_slack_report [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    count = counts[group_name] || 0
+    next nil if count == 0
+
+    url = Zendesk.group_url(group: group_name, statuses: STATUSES_TO_REPORT)
+    statuses = STATUSES_TO_REPORT.join('/')
+    "<#{url}|#{group_name} has #{count} #{statuses} Zendesk tickets.>"
+  end.compact
+
+  # Don't post if there's nothing to report
+  next if group_messages.empty?
```
Comment (islemaster) [inline bin/cron/zendesk_slack_report:135]: Note, we intentionally post _nothing_ if the count is zero for a channel.

I'd like to keep Slack notifications actionable, and zero tickets is not an actionable notification.  If we want this for metrics purposes, let's extend this script to log counts to redshift or something.

File: bin/cron/zendesk_slack_report [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  # "developers" => ["Developers"],
+  # "plc-engineering" => ["PLC Tools"],
+  # "privacy" => ["Privacy"],
+  # "user-accounts" => ["Accounts"],
+}
+
+# Zendesk group -> Zendesk filter id mappings for helpful links (optional)
+# rubocop:disable Style/NumericLiterals
+GROUP_FILTERS = {
+  "Developers" => 47222126,
```
Comment (davidsbailey) [inline bin/cron/zendesk_slack_report]: these values could just be strings, if you want to eliminate the rubocop warning

**Review by davidsbailey — APPROVED** [review]
Very nice, Brad!

**Comment by dmcavoy** [comment]
Looks really cool Brad! I don't feel confident giving a review on this material but I appreciate you including me so I could look through it and learn some new things!

**Review by maddiedierker — APPROVED** [review]
🎉 this change looks great, and i love the documentation!

**Comment by islemaster** [comment]
Update: Almost four months later, I'd say this little script has been a big success.  Although it's not totally clear that this reminder is the cause, our team's doing much better at managing our Zendesk backlog.  The big star here is the StarLabs team, who get a huge volume of requests and have caught up and kept up!

image (11)

**Pull request merged**